yarn-project/noir-contracts/src/contracts/noir-aztec3/src/public_call_stack_item.nr

```diff
@@ -0,0 +1,97 @@
+
+#[oracle(enqueuePublicFunctionCall)]
+fn enqueue_public_function_call_oracle(
+    _contract_address: Field, 
+    _function_selector: Field, 
+    _args: [Field; crate::abi::MAX_ARGS]
+) -> [Field; ENQUEUE_PUBLIC_FUNCTION_CALL_RETURN_SIZE] {}
+
+unconstrained fn enqueue_public_function_call_internal(
+    contract_address: Field, 
+    function_selector: Field,
+    args: [Field; crate::abi::MAX_ARGS]
+) -> [Field; ENQUEUE_PUBLIC_FUNCTION_CALL_RETURN_SIZE] {
+    enqueue_public_function_call_oracle(
+        contract_address, 
+        function_selector, 
+        args,
+    )
+}
+
+struct PublicCallStackItem {
+    contract_address: Field,
+    function_data: FunctionData,
+    public_inputs: PublicCircuitPublicInputs,
+}
+
+impl PublicCallStackItem {
+    fn call(contract_address: Field, function_selector: Field, args: [Field; crate::abi::MAX_ARGS], mut context: PrivateFunctionContext) -> (Self, PrivateFunctionContext) {
+        let fields = enqueue_public_function_call_internal(
+            contract_address, 
+            function_selector, 
+            args
+        );
+        let item = PublicCallStackItem {
+            contract_address: fields[0],
+            function_data: FunctionData {
+                function_selector: fields[1],
+                is_private: fields[2] as bool,
+                is_constructor: fields[3] as bool,
+            },
+            public_inputs: PublicCircuitPublicInputs {
+                call_context: CallContext {
+                    msg_sender : fields[4],
+                    storage_contract_address : fields[5],
+                    portal_contract_address : fields[6],
+
+                    is_delegate_call : fields[7] as bool,
+                    is_static_call : fields[8] as bool,
+                    is_contract_deployment: fields[9] as bool
+                },
+                args: arr_copy_slice(fields, [0; crate::abi::MAX_ARGS], 10),
+                return_values: [0; crate::abi::MAX_RETURN_VALUES],
+                emitted_events: [0; crate::abi::MAX_LOGS],
+                contract_storage_update_requests: [crate::abi::empty_contract_storage_update_request(); crate::abi::CONTRACT_STORAGE_UPDATE_REQUESTS_LENGTH],
+                contract_storage_read: [crate::abi::empty_contract_storage_read(); crate::abi::CONTRACT_STORAGE_READS_LENGTH],
+                public_call_stack: [0; crate::abi::MAX_PUBLIC_CALL_STACK],
+                new_l2_to_l1_msgs:[0; crate::abi::MAX_L1_MSG_STACK],
+                historic_public_data_tree_root: 0,
+                prover_address: 0,
+            },
+        };
+        constrain contract_address == item.contract_address;
+        constrain function_selector == item.function_data.function_selector;
+        for i in 0..crate::abi::MAX_ARGS {
+            constrain args[i] == item.public_inputs.args[i];
+        }
+
+        // TODO: Do we need to constrain the returned call_context here?
```

Comment:
@sirasistant heads up about this TODO when you review the PR

Reply:
I think we should probably constrain the `call_context` in both the kernel circuit _and_ in this `call` function. Just to be safe.
I _think_ you're right, Alvaro, that they're not needed. But equality checks are so cheap, I'd rather _over_-constrain, for now. Relying on a check to be done in the kernel circuit is dangerous, in case someone removes that check at some point in the future.